pyqg/model.py

```diff
@@ -550,6 +552,24 @@ def _step_forward(self):
 
         # the basic steps are
         self._print_status()
+    
+    def _invert(self):
+        self.kernel.invert()
+
+    def _do_advection(self):
+        self.kernel.do_advection()
+
+    def _do_friction(self):
+        self.kernel.do_friction()
+
+    def _do_q_subgrid_parameterization(self):
+        self.kernel.do_q_subgrid_parameterization()
+
+    def _do_uv_subgrid_parameterization(self):
+        self.kernel.do_uv_subgrid_parameterization()
+
+    def _forward_timestep(self):
+        self.kernel.forward_timestep()
```

Comment:
This feels like a code smell. Can we avoid wrapping all these methods? Could we use delegate here?

Reply:
I agree that's better. The issue is that, having removed underscores from the corresponding `kernel` method names, then we end up needing to remove underscores from the `pyqg.Model` method names as well, which means they're now part of the public API (and it also breaks some of my code which calls the private methods anyway 😬).

A terse and backwards-compatible alternative I spent a little while starting to implement was updating `delegate` to support a prefix option:

```python
@delegate(*private_kernel_attrs, to='kernel', prefix='_')
@delegate(*public_kernel_attrs, to='kernel')
class Model:
    # ...
```

This is relatively simple in principle, though it requires significant changes to the `delegate` decorator to allow calls to stack / prevent one call from overwriting the other. Because of that, I opted for the smelly solution. However, with your comment, I'm considering updating the decorator again...

Reply:
Alright, I updated it, and also made a pull request to the original library in case they want to allow this functionality (https://github.com/dscottboggs/python-delegate)

Reply:
From what I can see, I think this can be resolved by composing most of the kernel functions. I think `Kernel._forward_timestep` is the only one that couldn't be, atm, but only because `_calc_diagnostics` is in the way.

I'm really wary of solving the Kernel/Model inheritance problem with delegation, it takes away a lot of the guarantees I need to be able to implement Jax #241 in a reasonable way. Basically Kernel would become part of Model, which means I can't know if the Kernel arrays are being messed with outside of Kernel. I think if we compose all the kernel functions, that would alleviate this problem on my end, though.